scripts/genererate_drivers.py

```diff
@@ -0,0 +1,633 @@
+    """
+    template = string.Template(textwrap.dedent(METAL_DEV_DRV_TMPL))
+
+    return template.substitute(
+        vendor=vendor,
+        device=device,
+        cap_device=device.upper(),
+        index=str(index),
+        offsets=generate_offsets(device, reglist),
+        base_functions=generate_base_functions(device, reglist),
+        metal_functions=generate_metal_function(device, reglist),
+        def_vtable=generate_def_vtable(device, reglist)
+    )
+
+
+###
+# main
+###
+
+
+def handle_args():
+    """
+    :return:
+    """
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
+        "-o",
+        "--object-model",
+        type=argparse.FileType('r'),
+        help="",
+    )
+
+    parser.add_argument(
+        "-d",
+        "--duh-file",
+        type=argparse.FileType('r'),
+        help="",
+    )
+
+    parser.add_argument(
+        "-v",
+        "--vendor",
+        help="",
+    )
+
+    parser.add_argument(
+        "-D",
+        "--device",
+        help="",
+    )
+
+    parser.add_argument(
+        "-m",
+        "--metal-dir",
+        help="",
+    )
+
+    parser.add_argument(
+        "-x",
+        "--check-existence",
+        action="store_true",
+        default=False
+    )
+
+    parser.add_argument(
+        "-H",
+        "--base-header-only",
+        action="store_true",
+        default=False
+    )
+
+    return parser.parse_args()
+
+
+def main():
+    """
+
+    :return: exits 0 on success, 1 on failure
+    """
+
+    # debugging
+    print(sys.argv, file=sys.stderr)
+
+    # parse args
+    args = handle_args()
+
+    object_model = json.load(args.object_model)
+    duh_info = json5.load(args.duh_file)
+    vendor = args.vendor
+    device = args.device
+    metal_dir = args.metal_dir
+    check_existence = args.check_existence
+    base_header_only = args.base_header_only
+    m_dir_path = Path(metal_dir)
+
+    # process register info from duh
+    addr_blocks = (i['addressBlocks'][0] for i in
+                   find_component(duh_info, 'addressBlocks'))
+
+    regs = next(addr_blocks)['registers']
+
+    reglist: t.List[Register] = []
+    for a_reg in regs:
+        name: str = a_reg['name']
+        offset: int = a_reg['addressOffset'] // 8
+        width: int = a_reg['size']
+        if width not in (8, 16, 32, 64):
+            raise Exception(f'Invalid Register width {width}, register: '
+                            f'{name}, not 8, 16, 32, or 64'
+                            f'fix register width in duh file')
+
+        reglist.append(Register(name, offset, width))
+
+    # parse OM to find base address of all devices
+    memory_regions = find_json_field_name(object_model, '_types', 'OMMemoryRegion')
+    devices = find_json_field_name(memory_regions, 'name', f'{device}.*')
+    devices_enumerated = list(enumerate(devices))
+
+    m_hdr_path = m_dir_path / device
+
+    if not m_hdr_path.exists():
+        m_hdr_path.mkdir()
```

Comment:
You can unconditionally call [`.mkdir(exist_ok=True)`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir) to avoid having to put the guard as a separate if statement.

Reply:
fixed